language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java

```diff
@@ -0,0 +1,443 @@
+      return value;
+    };
+  }
+
+  public FromJson<Instant> timestamp() {
+    return () -> {
+      expectIsAt("timestamp", JsonToken.VALUE_STRING);
+      Instant value = null;
+      try {
+        value = Instant.parse(parser.getText());
+      } catch (DateTimeParseException e) {
+        parseExpected("timestamp");
+      } catch (IOException e) {
+        parseExpected("timestamp");
+      }
+      moveNext();
+      return value;
+    };
+  }
+
+  public FromJson<LocalDate> date() {
+    return () -> {
+      expectIsAt("date", JsonToken.VALUE_STRING);
+      LocalDate value = null;
+      try {
+        value = LocalDate.parse(parser.getText());
+      } catch (DateTimeParseException e) {
+        parseExpected("date");
+      } catch (IOException e) {
+        parseExpected("date");
+      }
+      moveNext();
+      return value;
+    };
+  }
+
+  public FromJson<String> party() {
+    return text();
+  }
+
+  public <C extends ContractId<?>> FromJson<C> contractId(Function<String, C> constr) {
+    return () -> {
+      String id = text().read();
+      return constr.apply(id);
+    };
+  }
+
+  public FromJson<String> text() {
+    return () -> {
+      expectIsAt("text", JsonToken.VALUE_STRING);
+      String value = null;
+      try {
+        value = parser.getText();
+      } catch (IOException e) {
+        parseExpected("valid textual value");
+      }
+      moveNext();
+      return value;
+    };
+  }
+
+  // Read an list with an unknown number of items of the same type.
+  public <T> FromJson<List<T>> list(FromJson<T> readItem) {
+    return () -> {
+      List<T> list = new ArrayList<>();
+      readStartArray();
+      while (notEndArray()) {
+        T item = readItem.read();
+        list.add(item);
+      }
+      readEndArray();
+      return list;
+    };
+  }
+
+  // Read a map with textual keys, and unknown number of items of the same type.
+  public <V> FromJson<Map<String, V>> textMap(FromJson<V> readValue) {
+    return () -> {
+      Map<String, V> map = new TreeMap<>();
+      readStartObject();
+      while (notEndObject()) {
+        String key = readFieldName();
+        V val = readValue.read();
+        map.put(key, val);
+      }
+      ;
+      readEndObject();
+      return map;
+    };
+  }
+
+  // Read a map with unknown number of items of the same types.
+  public <K, V> FromJson<Map<K, V>> genMap(FromJson<K> readKey, FromJson<V> readValue) {
+    return () -> {
+      Map<K, V> map = new TreeMap<>();
+      // Maps are represented as an array of 2-element arrays.
+      readStartArray();
+      while (notEndArray()) {
+        readStartArray();
+        K key = readKey.read();
+        V val = readValue.read();
+        readEndArray();
+        map.put(key, val);
+      }
+      ;
+      readEndArray();
+      return map;
+    };
+  }
+
+  // The T type should not itself be Optional<?>. In that case use OptionalNested below.
+  public <T> FromJson<Optional<T>> optional(FromJson<T> readValue) {
```

Comment:
is it worth adding a runtime typecheck that T is not an instance of Optional?

Reply:
Sure. If this happened, the bug would be in the code-gen, so they're an extra step removed from being able to handle it, but at least we can report it in a useful way.

I'd actually like to come back to this to see whether we can avoid having these two different decoders but this was what I got working initially.

But in the meantime, done.